File: library-core/src/main/java/com/mikepenz/iconics/context/InternalLayoutInflater.java
package com.mikepenz.iconics.context;

import android.annotation.TargetApi;
import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.xmlpull.v1.XmlPullParser;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

/**
 * Base created by Christopher Jenkins
 * https://github.com/chrisjenx/Calligraphy
 */
class InternalLayoutInflater extends LayoutInflater {

    private static final String[] sClassPrefixList = {
            "android.widget.",
            "android.webkit."
    };

    private final IconicsFactory mIconicsFactory;
    // Reflection Hax
    private boolean mSetPrivateFactory = false;
    private Field mConstructorArgs = null;

    protected InternalLayoutInflater(Context context) {
        super(context);
        mIconicsFactory = new IconicsFactory();
        setUpLayoutFactories(false);
    }

    protected InternalLayoutInflater(LayoutInflater original, Context newContext, boolean cloned) {
        super(original, newContext);
        mIconicsFactory = new IconicsFactory();
        setUpLayoutFactories(cloned);
    }

    @Override
    public LayoutInflater cloneInContext(Context newContext) {
        return new InternalLayoutInflater(this, newContext, true);
    }

    @Override
    protected View onCreateView(String name, AttributeSet attrs) throws ClassNotFoundException {
        View view = null;
        for (String prefix : sClassPrefixList) {
            try {
                view = createView(name, prefix, attrs);
            } catch (ClassNotFoundException ignored) {
            }
        }
        // In this case we want to let the base class take a crack
        // at it.
        if (view == null) {
            view = super.onCreateView(name, attrs);
        }

        return mIconicsFactory.onViewCreated(view, view.getContext(), attrs);
    }

    @Override
    protected View onCreateView(View parent, String name, AttributeSet attrs) throws ClassNotFoundException {
        return mIconicsFactory.onViewCreated(
                super.onCreateView(parent, name, attrs),
                getContext(), attrs);
    }

    // ===
    // Wrapping goodies
    // ===

    @Override
    public View inflate(XmlPullParser parser, ViewGroup root, boolean attachToRoot) {
        setPrivateFactoryInternal();
        return super.inflate(parser, root, attachToRoot);
    }

    /**
     * We don't want to unnecessary create/set our factories if there are none there. We try to be
     * as lazy as possible.
     */
    private void setUpLayoutFactories(boolean cloned) {
        if (cloned) {
            return;
        }
        if (getFactory2() != null && !(getFactory2() instanceof WrapperFactory2)) {
            // Sets both Factory/Factory2
            setFactory2(getFactory2());
        }

    }

    @Override
    public void setFactory(Factory factory) {
        // Only set our factory and wrap calls to the Factory trying to be set!
        if (!(factory instanceof WrapperFactory)) {
            super.setFactory(new WrapperFactory(factory, this, mIconicsFactory));
        } else {
            super.setFactory(factory);
        }
    }

    @Override
    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    public void setFactory2(Factory2 factory2) {
        // Only set our factory and wrap calls to the Factory2 trying to be set!
        if (!(factory2 instanceof WrapperFactory2)) {
//            LayoutInflaterCompat.setFactory(this, new WrapperFactory2(factory2, mCalligraphyFactory));
            super.setFactory2(new WrapperFactory2(factory2, mIconicsFactory));
        } else {
            super.setFactory2(factory2);
        }
    }

    private void setPrivateFactoryInternal() {
        // Already tried to set the factory.
        if (mSetPrivateFactory) {
            return;
        }

        // Skip if not attached to an activity.
        if (!(getContext() instanceof Factory2)) {
            mSetPrivateFactory = true;
            return;
        }

        final Method setPrivateFactoryMethod = ReflectionUtils
                .getMethod(LayoutInflater.class, "setPrivateFactory");

        if (setPrivateFactoryMethod != null) {
            ReflectionUtils.invokeMethod(
                    this,
                    setPrivateFactoryMethod,
                    new PrivateWrapperFactory2((Factory2) getContext(), this, mIconicsFactory));
        }
        mSetPrivateFactory = true;
    }

    private View createCustomViewInternal(View parent, View view, String name, Context viewContext, AttributeSet attrs) {
        boolean customViewCreation = true;
        if (!customViewCreation) {
            return view;
        }
        if (view == null && name.indexOf('.') > -1) {
            if (mConstructorArgs == null) {
                mConstructorArgs = ReflectionUtils.getField(LayoutInflater.class, "mConstructorArgs");
            }

            final Object[] mConstructorArgsArr = (Object[]) ReflectionUtils.getValue(mConstructorArgs, this);
            final Object lastContext = mConstructorArgsArr[0];
            // The LayoutInflater actually finds out the correct context to use. We just need to set
            // it on the mConstructor for the internal method.
            // Set the constructor ars up for the createView, not sure why we can't pass these in.
            mConstructorArgsArr[0] = viewContext;
            ReflectionUtils.setValue(mConstructorArgs, this, mConstructorArgsArr);
            try {
                view = createView(name, null, attrs);
            } catch (ClassNotFoundException ignored) {
            } finally {
                mConstructorArgsArr[0] = lastContext;
                ReflectionUtils.setValue(mConstructorArgs, this, mConstructorArgsArr);
            }
        }
        return view;
    }

    // ===
    // Wrapper Factories for Pre/Post HC
    // ===

    /**
     * Factory 1 is the first port of call for LayoutInflation
     */
    private static class WrapperFactory implements Factory {

        private final Factory mFactory;
        private final InternalLayoutInflater mInflater;
        private final IconicsFactory mIconicsFactory;

        public WrapperFactory(Factory factory, InternalLayoutInflater inflater, IconicsFactory iconicsFactory) {
            mFactory = factory;
            mInflater = inflater;
            mIconicsFactory = iconicsFactory;
        }

        @Override
        public View onCreateView(String name, Context context, AttributeSet attrs) {
            return mIconicsFactory.onViewCreated(
                    mFactory.onCreateView(name, context, attrs),
                    context, attrs
            );
        }
    }

    /**
     * Factory 2 is the second port of call for LayoutInflation
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    private static class WrapperFactory2 implements Factory2 {
        protected final Factory2 mFactory2;
        protected final IconicsFactory mIconicsFactory;

        public WrapperFactory2(Factory2 factory2, IconicsFactory iconicsFactory) {
            mFactory2 = factory2;
            mIconicsFactory = iconicsFactory;
        }

        @Override
        public View onCreateView(String name, Context context, AttributeSet attrs) {
            return mIconicsFactory.onViewCreated(
                    mFactory2.onCreateView(name, context, attrs),
                    context, attrs);
        }

        @Override
        public View onCreateView(View parent, String name, Context context, AttributeSet attrs) {
            return mIconicsFactory.onViewCreated(
                    mFactory2.onCreateView(parent, name, context, attrs),
                    context, attrs);
        }
    }

    /**
     * Private factory is step three for Activity Inflation, this is what is attached to the
     * Activity on HC+ devices.
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    private static class PrivateWrapperFactory2 extends WrapperFactory2 {

        private final InternalLayoutInflater mInflater;

        public PrivateWrapperFactory2(Factory2 factory2, InternalLayoutInflater inflater, IconicsFactory iconicsFactory) {
            super(factory2, iconicsFactory);
            mInflater = inflater;
        }

        @Override
        public View onCreateView(View parent, String name, Context context, AttributeSet attrs) {
            return mIconicsFactory.onViewCreated(
                    mInflater.createCustomViewInternal(
                            parent,
                            mFactory2.onCreateView(parent, name, context, attrs),
                            name, context, attrs
                    ),
                    context, attrs
            );
        }
    }
}

File: library-core/src/main/java/com/mikepenz/iconics/context/InternalLayoutInflater.kt
/*
 * Copyright 2019 Mike Penz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mikepenz.iconics.context

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import org.xmlpull.v1.XmlPullParser
import java.lang.reflect.Field

/**
 * Base created by Christopher Jenkins
 * https://github.com/chrisjenx/Calligraphy
 */
internal class InternalLayoutInflater : LayoutInflater {
    // Reflection Hax
    private var isSetPrivateFactory = false
    private var constructorArgs: Field? = null

    protected constructor(context: Context) : super(context) {
        setUpLayoutFactories(false)
    }

    constructor(
        original: LayoutInflater,
        newContext: Context,
        cloned: Boolean
    ) : super(original, newContext) {
        setUpLayoutFactories(cloned)
    }

    override fun cloneInContext(newContext: Context): LayoutInflater {
        return InternalLayoutInflater(this, newContext, true)
    }

    @Throws(ClassNotFoundException::class)
    override fun onCreateView(name: String, attrs: AttributeSet): View? {
        var view: View? = null

        classPrefixList.forEach { runCatching { view = createView(name, it, attrs) } }
        // In this case we want to let the base class take a crack at it.
        if (view == null) {
            view = super.onCreateView(name, attrs)
        }

        return view?.let { IconicsFactory.onViewCreated(it, it.context, attrs) }
    }

    @Throws(ClassNotFoundException::class)
    override fun onCreateView(parent: View?, name: String, attrs: AttributeSet): View? {
        val view = super.onCreateView(parent, name, attrs)
        return IconicsFactory.onViewCreated(view, context, attrs)
    }

    // ===
    // Wrapping goodies
    // ===

    override fun inflate(parser: XmlPullParser, root: ViewGroup?, attachToRoot: Boolean): View {
        setPrivateFactoryInternal()
        return super.inflate(parser, root, attachToRoot)
    }

    /**
     * We don't want to unnecessary create/set our factories if there are none there. We try to be
     * as lazy as possible.
     */
    private fun setUpLayoutFactories(cloned: Boolean) {
        if (cloned) {
            return
        }
        if (factory2 != null && factory2 !is InternalLayoutInflater.WrapperFactory2) {
            // Sets both Factory/Factory2
            factory2 = factory2
        }
    }

    override fun setFactory(factory: LayoutInflater.Factory) {
        // Only set our factory and wrap calls to the Factory trying to be set!
        if (factory !is InternalLayoutInflater.WrapperFactory) {
            super.setFactory(InternalLayoutInflater.WrapperFactory(factory))
        } else {
            super.setFactory(factory)
        }
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    override fun setFactory2(factory2: LayoutInflater.Factory2) {
        // Only set our factory and wrap calls to the Factory2 trying to be set!
        if (factory2 !is InternalLayoutInflater.WrapperFactory2) {
            super.setFactory2(InternalLayoutInflater.WrapperFactory2(factory2))
        } else {
            super.setFactory2(factory2)
        }
    }

    private fun setPrivateFactoryInternal() {
        // Already tried to set the factory.
        if (isSetPrivateFactory) {
            return
        }

        // Skip if not attached to an activity.
        if (context !is LayoutInflater.Factory2) {
            isSetPrivateFactory = true
            return
        }

        val setPrivateFactoryMethod = ReflectionUtils.getMethod(
            LayoutInflater::class.java,
            "setPrivateFactory"
        )

        if (setPrivateFactoryMethod != null) {
            ReflectionUtils.invokeMethod(
                this,
                setPrivateFactoryMethod,
                PrivateWrapperFactory2(
                    context as LayoutInflater.Factory2,
                    this
                )
            )
        }
        isSetPrivateFactory = true
    }

    private fun createCustomViewInternal(
        view: View?,
        name: String,
        viewContext: Context,
        attrs: AttributeSet
    ): View? {
        var createdView = view
        if (createdView == null && name.contains('.')) {
            if (constructorArgs == null) {
                constructorArgs = ReflectionUtils.getField(
                    LayoutInflater::class.java,
                    "mConstructorArgs"
                )
            }
            val constructorArgs = this.constructorArgs ?: return null

            @Suppress("UNCHECKED_CAST")
            val constArgs = ReflectionUtils.getValue(constructorArgs, this) as Array<Any>
            val lastContext = constArgs[0]
            // The LayoutInflater actually finds out the correct context to use. We just need to set
            // it on the mConstructor for the internal method.
            // Set the constructor ars up for the createView, not sure why we can't pass these in.
            constArgs[0] = viewContext
            ReflectionUtils.setValue(constructorArgs, this, constArgs)

            runCatching { createdView = createView(name, null, attrs) }
            constArgs[0] = lastContext
            ReflectionUtils.setValue(constructorArgs, this, constArgs)
        }
        return createdView
    }

    // ===
    // Wrapper Factories for Pre/Post HC
    // ===

    /** Factory 1 is the first port of call for LayoutInflation */
    private class WrapperFactory(
        private val factory: LayoutInflater.Factory
    ) : LayoutInflater.Factory {

        override fun onCreateView(name: String, context: Context, attrs: AttributeSet): View? {
            val view = factory.onCreateView(name, context, attrs)
            return IconicsFactory.onViewCreated(view, context, attrs)
        }
    }

    /** Factory 2 is the second port of call for LayoutInflation */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    private open class WrapperFactory2(
        protected val factory2: LayoutInflater.Factory2
    ) : LayoutInflater.Factory2 {

        override fun onCreateView(name: String, context: Context, attrs: AttributeSet): View? {
            val view = factory2.onCreateView(name, context, attrs)
            return IconicsFactory.onViewCreated(view, context, attrs)
        }

        override fun onCreateView(
            parent: View?,
            name: String,
            context: Context,
            attrs: AttributeSet
        ): View? {
            val view = factory2.onCreateView(parent, name, context, attrs)
            return IconicsFactory.onViewCreated(view, context, attrs)
        }
    }

    /**
     * Private factory is step three for Activity Inflation, this is what is attached to the
     * Activity on HC+ devices.
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    private class PrivateWrapperFactory2(
        factory2: LayoutInflater.Factory2,
        private val inflater: InternalLayoutInflater
    ) : InternalLayoutInflater.WrapperFactory2(factory2) {

        override fun onCreateView(
            parent: View?,
            name: String,
            context: Context,
            attrs: AttributeSet
        ): View? {
            val view = factory2.onCreateView(parent, name, context, attrs)
            val customView = inflater.createCustomViewInternal(view, name, context, attrs)
            return IconicsFactory.onViewCreated(customView, context, attrs)
        }
    }

    companion object {
        private val classPrefixList = arrayOf("android.widget.", "android.webkit.")
    }
}